File: app/src/main/java/com/example/nijimac103/easyclient_mvvm/view/ui/ProjectFragment.java
package com.example.nijimac103.easyclient_mvvm.view.ui;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.example.nijimac103.easyclient_mvvm.R;
import com.example.nijimac103.easyclient_mvvm.databinding.FragmentProjectDetailsBinding;
import com.example.nijimac103.easyclient_mvvm.service.model.Project;
import com.example.nijimac103.easyclient_mvvm.viewModel.ProjectViewModel;

import java.util.Objects;


public class ProjectFragment extends Fragment {

    private static final String KEY_PROJECT_ID = "project_id";
    private FragmentProjectDetailsBinding binding;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        // DataBinding対象のレイアウトをinflateする
        binding = DataBindingUtil.<FragmentProjectDetailsBinding>inflate(inflater, R.layout.fragment_project_details, container, false);

        // Create and set the adapter for the RecyclerView.
        return binding.getRoot();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        //DI
        assert getArguments() != null;
        ProjectViewModel.Factory factory = new ProjectViewModel.Factory(
                Objects.requireNonNull(getActivity()).getApplication(), getArguments().getString(KEY_PROJECT_ID)
        );

        //project_idをキーに注入してViewModelインスタンスを取得
        final ProjectViewModel viewModel = ViewModelProviders.of(this, factory).get(ProjectViewModel.class);

        //ViewにViewModelをセット
        binding.setProjectViewModel(viewModel);
        //app:visibleGone="@{isLoading}"をtrueに
        binding.setIsLoading(true);

        //データ監視を開始 -> 差分を監視して、ViewModelに伝える
        observeViewModel(viewModel);

    }

    //Modelのデータを監視するメソッド
    public void observeViewModel(final ProjectViewModel viewModel){
       viewModel.getObservableProject().observe(this, new Observer<Project>() {
           @Override
           public void onChanged(@Nullable Project project) {
               if (project != null){

                   binding.setIsLoading(false);

                   viewModel.setProject(project);
               }
           }
       });
    }


    //fragmentTofragmentでidを渡す
    public static ProjectFragment forProject(String projectID) {
        ProjectFragment fragment = new ProjectFragment();
        Bundle args = new Bundle();

        args.putString(KEY_PROJECT_ID, projectID);
        fragment.setArguments(args);

        return fragment;
    }

}

File: app/src/main/java/jp/sample/tsutou/githubClient/view/ui/ProjectFragment.kt
package jp.sample.tsutou.githubClient.view.ui

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import jp.sample.tsutou.githubClient.R
import jp.sample.tsutou.githubClient.databinding.FragmentProjectDetailsBinding
import jp.sample.tsutou.githubClient.viewModel.ProjectViewModel

import java.util.Objects

private const val KEY_PROJECT_ID = "project_id"

class ProjectFragment : Fragment() {

    private var binding: FragmentProjectDetailsBinding? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // DataBinding対象のレイアウトをinflateする
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_project_details, container, false)

        // Create and set the adapter for the RecyclerView.
        return requireNotNull(this.binding).root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        //DI
        assert(arguments != null)
        val factory = ProjectViewModel.Factory(
                Objects.requireNonNull<FragmentActivity>(activity).application, requireNotNull(arguments).getString(KEY_PROJECT_ID)!!
        )

        //project_idをキーに注入してViewModelインスタンスを取得
        val viewModel = ViewModelProviders.of(this, factory).get(ProjectViewModel::class.java)

        //ViewにViewModelをセット
        requireNotNull(binding).projectViewModel = viewModel

        //app:visibleGone="@{isLoading}"をtrueに
        requireNotNull(binding).isLoading = true

        //データ監視を開始 -> 差分を監視して、ViewModelに伝える
        observeViewModel(viewModel)

    }

    //Modelのデータを監視するメソッド
    private fun observeViewModel(viewModel: ProjectViewModel) {
        viewModel.observableProject.observe(this, Observer { project ->
            if (project != null) {

                requireNotNull(binding).isLoading = false

                viewModel.setProject(project)
            }
        })
    }

    companion object PUT {

        //fragmentToFragmentでidを渡す
        fun forProject(projectID: String): ProjectFragment {
            val fragment = ProjectFragment()
            val args = Bundle()

            args.putString(KEY_PROJECT_ID, projectID)
            fragment.arguments = args

            return fragment
        }
    }

}
